### ISSUE TITLE
SkillTree Generator Crashes with Null Exception

### ISSUE THREAD

**Issue by EmmittJ** [EmmittJ/PoESkillTree#321]
@brather1ng line 325 of `AdvancedSolver.cs` with have a null exception if you check tag any node, and I am not sure how to fix this issue.

**Comment by brather1ng** [comment]
I'll look into it when I'm home. Which is in like 8 hours.

**Comment by EmmittJ** [comment]
No worries

**Comment by EmmittJ** [comment]
![fix](http://i.imgur.com/B9cIniV.png)
That fix doesn't exactly work. If you select two tags now then it will throw a `GraphNotConnectedException`

**Comment by brather1ng** [comment]
> If you select two tags

What do you mean with that? I can get it to look like in your picture (which obviously is a bug), but I can't get it throw something.

**Comment by EmmittJ** [comment]
My apologizes, I tried to skill Phase Acro without skilling Acro. 

I am not sure if this is related, but now some extra nodes get skilled during the 'Skill tagged node' option. 
![extra nodes](http://i.imgur.com/cASeowr.jpg)

**Comment by brather1ng** [comment]
They are both unrelated to the first one.

The first one only seems to only happen if no constraints are selected, in which case the Skill Tree Generator is not the best option anyway. It doesn't seem to hard to fix, I'll hopefully get to fix it in the next few hours.

The second one is probably a bug in one of the reduction tests. That may take a while to track down.

**Comment by EmmittJ** [comment]
Ah, sorry! Thanks for the work. I would try and help, but I don't really
understand what is going on. I also have class in a bit, so I won't be at a
computer

**Comment by brather1ng** [comment]
I fixed the first bug. I might be able to fix the second one fast, I might not, but it should only happen in very few cases so it shouldn't stop you from doing a release if you want to do that.

**Comment by EmmittJ** [comment]
Yep, that was the plan, and I was just running through all the features we
have. I am trying to test more stuff before releasing, so we can catch
stuff like this.

**Comment by brather1ng** [comment]
We should really have more unit tests. The problem is, most of our code isn't testable, everything depends on everything and the GUI stuff is woven into everything too. At some point we'll probably have to refactor most of our code.

**Comment by EmmittJ** [comment]
I also doesn't help that we have stuff that relies on, for instance, the
stats gems actually have instead of fake stats. It would make things easier
if we were to make ourselves some test data like a tree, images, builds,
etc.

**Comment by brather1ng** [comment]
Yeah, hardcoding test input and output as a bunch of fake numbers is not a good idea. I don't know what we do with those calculation unit tests but it's probably just changing the expected output to match the actual output every time they fail because no one wants to actually calculate all of it.

Changing all of those existing unit tests, refactoring the whole code to be properly unit testable and writing a bunch of new tests would be quite a big undertaking.

**Comment by MLanghof** [comment]
I tried to make unit tests for some parts of the steiner problem back in the day, but I didn't end up building any tool that would facilitate creating trees. And being stuck with manually writing adjacency matrices (or writing a line of code for every connection) means it's really tedious to create anything larger than 8-10 node trees (n² entries, yay), which just doesn't cut it for actually producing reliable test cases. Plus you're very likely to screw up the input.

Anything that relies on GGG's tree data instead is also iffy, because any part of the tree might change, and we probably don't want to keep shipping older trees with the unit tests because they may or may not break.

**Comment by brather1ng** [comment]
I extended your unit tests a bit but not by much because, as you said, creating test trees is not fun, and all the classes that are not bottom layer already have skill tree specific stuff in them, which makes testing it hard because it would need to load the whole SkillTree stuff.

>  Already, SkillTree.cs and SkillTree-Drawing.cs are monstrous (and they're the same class).

Making actually different classes out of them would be a first step. Rewriting MainWindow (it's also monstrous) first should be easier, though. It shouldn't be too hard to make individual controls out of it and refactor those into proper MVVM. For an example of how that looks, all of the TreeGenerator UI stuff is done with that architecture. It's not really difficult, I didn't know anything about it before this.

This went a bit off-topic. If someone wants to actually start on that, it should be discussed in a separate issue, I guess.

@EmmittJ I have tracked down the last bug to the point where I know what causes it and what should fix it, but I have to go to bed now. I'll fix that tomorrow. It probably is somewhat rare and seems to only impact problems where it is obvious to the user if it occurs (like in your example).
